sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriterV2.scala

```diff
@@ -104,9 +106,27 @@ final class DataFrameWriterV2[T] private[sql](table: String, ds: Dataset[T])
     }
 
     this.partitioning = Some(asTransforms)
+    validatePartitioning()
+    this
+  }
+
+  @scala.annotation.varargs
+  override def clusterBy(colName: String, colNames: String*): CreateTableWriter[T] = {
+    this.clustering =
+      Some(ClusterByTransform((colName +: colNames).map(col => FieldReference(col))))
+    validatePartitioning()
     this
   }
 
+  /**
+   * Validate that clusterBy is not used with partitionBy.
+   */
+  private def validatePartitioning(): Unit = {
+    if (partitioning.nonEmpty && clustering.nonEmpty) {
```

Comment:
`DataFrameWriterV2` does not have `bucketBy`?

Reply:
For DFW V2, [bucketBy](https://github.com/apache/spark/blob/3af2ea973c458b9bd9818d6af733683fb15fbc19/sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriterV2.scala#L98) is implemented as a special partitionedBy transform, therefore I didn't call out here.